### PR TITLE
Minimize memory footprint for AbstractChannelHandlerContext for handl…

### PR DESCRIPTION
**Pull request by normanmaurer**
…ers that execute in the EventExecutor.

Motivation:

We cache the Runnable for some tasks to reduce GC pressure in 4 different fields. This gives overhead in terms of memory usage in all cases, even if we always execute in the EventExecutor (which is the case most of the times).

Modifications:

Move the 4 fields to another class and only have one reference to this in AbstractChannelHandlerContext. This gives a small overhead in the case of execution that is done outside of the EventExecutor but reduce memory footprint in the more likily execution case.

Result:

Less memory used per AbstractChannelHandlerContext in most cases.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@njhill @franz1981

**Comment by normanmaurer** [comment]
@netty-bot Test this please

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 1]
```diff
@@ -1162,4 +1139,36 @@ public void write(AbstractChannelHandlerContext ctx, Object msg, ChannelPromise
             ctx.invokeFlush();
         }
     }
+
+    private static final class Tasks {
+        private final AbstractChannelHandlerContext next;
+        private final Runnable invokeChannelReadCompleteTask = new Runnable() {
```
Comment (franz1981) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: I was expecting this one to be initailized to `null`!

Comment (franz1981) [reply transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Now I've understood the logic :) it's fine!

File: transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java [hunk 2]
```diff
@@ -1162,4 +1139,36 @@ public void write(AbstractChannelHandlerContext ctx, Object msg, ChannelPromise
...
+            @Override
+            public void run() {
+                next.invokeChannelReadComplete();
+            }
+        };
+        private Runnable invokeReadTask = new Runnable() {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java]: Doh... all of these should be final. Will fix later today

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Pull request merged**